### DISCUSSION TITLE
Allow spread operators to work on tuple types

### DISCUSSION THREAD

**Discussion by Captain-Quack** [dotnet/csharplang#7522] [Language Ideas] [answered]
Since tuples have an ordered number of objects that can all be accessed through the `ItemN`  property, the spread operator could be extended to work on them:
```csharp
(int, int) a = (1, 2); 
int[] b = [..a, 3] 	// { 1, 2, 3 }
```
And in reverse...
```csharp
int[] b = [1,2]; 
(int, int) a = (..b); 
```

The only current issue preventing this from happening (I believe) is the lack of `GetEnumerator`.

**Comment by CyrusNajmabadi** [comment, marked as answer]
GetEnumerator can be provided by an extension method.  So you can make  `int[] b = [..a, 3]` work just by making tuples enumerable.

**Reply by siegfriedpammer** [reply]
@Captain-Quack see https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xAdAMID2ANmwKYDGCULMAM4BuCtTrEALKMoBmOoRpMaAbxoUamuvPoA2OpJoBZaDAAU9KgG0AujQjIA5oICUGrSvdatANwc0EVAAHLhoAXhozYjgaQhcZb29YBFsaNihBBHCaKwYGQJDOGNkbBMTNL0SAMxZkTghuAAtIv2QaKAROMHaYAOCuN3Jy8voATjMOrvjK7xq6huazVvbO7tg0jIRB4cSxidXpoeGZrQB6U6DkZMEaaNiYu8JiyoBfCjfyMR1ifSIaABV+px/ixWBweHwBABRAAenSE/CEFE8R00+G++gAklCYKgwJxkBAELUADz/AB8NAA4pwEDi8QSiaSKWYEI0MjQAGoQNioYFAskxCl9QrbDwnNHEYh0ADsIq4DExq2IZRGUtl8s4itWhFVNA+3kq5wBAHkACImmggGgYThVWCcGhgWqOzhwzgIgQ3Ob2Dg0Fhsgk0VpQCAwBA3FhVLk8vmAwrvIA= for a working example

**Reply by Captain-Quack** [reply]
Yeah, that's true.

**Reply by sharpchen** [reply]
Or you can do
```C#
public static IEnumerator<T> GetEnumerator<T>(this ITuple tuple)
{
    for (int i = 0; i < tuple.Length; i++)
        yield return (T)tuple[i]!;
}

```
It should work for all versions of tuple has only one element type.
```C#
// Works
foreach (var i in (1, 2, 3, 4, 5, 6, 7, 8, 9))
    Console.WriteLine(i);

// System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Int32'.
foreach (var i in (1, 2, 3, 4, 5, 6, 7, 8, 9, "I am the last element.")
    Console.WriteLine(i);

// System.InvalidCastException: Unable to cast object of type 'System.Byte' to type 'System.Int32'.
foreach(var i in (1, 2, 3 ,4, 5, 6, 7, 8, 9, (byte)10))
    Console.WriteLine(i);
```

**Reply by FaustVX** [reply]
@UPIMMUNITY 
How can the compiler knows what `T` is ?
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4ACATAIwCwAUGQAwAEZVALANz1OtUB0AJQCuMBFDCp+AYQD2YAA5QANhgDKGZFADGqAM7c69APTHmAdVnoA1nvpYrqCNoAWzABSwEzKD5geqOGYKIIBmIPYggFYggDYggHYggA4ggE4ASgz6Zly+NM8MwxMzNn4ASRhkCGUofGkIPQQAUQAPXQVxWRgQZgBVGAhgVWYEWWZtRu9ZYAArVG1prFGcBVRmAHIytQRMGABzDdHxhFX1rYFKhFCKDf57R2c3d2r0Hz8AoJDmcOZI5hizHizCSzFSzDSQQAROVmBAwKMXOtlFNmKhVJIxPwoVkcnk2AUoEV6CU+BUqjU6g0mm0Ol0ev1BsN1mMJqiZvNFsxZMtTmtNmUAEI4BCoI6svnnMpXG53B7oJyuF4QN6+WCfYJhZhwf6A4Gg8GQjzAEWoDJURi4uh5fKFQw2h08UJ8WKsCjMNoIegAbxJ1vxzrYrvKzRgwkk6AgY3QAB4ACoAPmYAHFUC0wxGo1Z4wn3AgXFA9MxynHhAoRggy6psv7cr7aw6HG9PGJ3gBeZiMTjvGOjKtSAAyqAO+e7UCIRBrDunfCorASHjjGUr5dQAG0oABdACE9ryAF96PugA==)

But with a non-generic `GetEnumerator` everything works
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xAdAMID2ANmwKYDGCULMAM4BuCtTqMASuj5hOzFmAAOULsgDKnZADco3TiIoUA9MZoB1FsgDWgigDMrnCNwAWNABSwENKL5iexHA0hMEAzMEALMEArMEAbMEA7MEAHMEAnACUWRQ0+RIZXlmi5CZm9AwAkjDaEGxQGEwQgggAogAe+kp8AiA0AKowEMBcNAgsNNwtPizAAFY8s/bjAJ5KnDQA5JXqCMiwAOZb45MI65s7jDUIYYRbDA5OLu5eMD5+sIHBoTQRNNEaHEaIkaCkaOkaBlggAiKo0CBgcauTZsGY0ThcOTvBgwnJ5Ar0IpQEpGcimCTVWr1RrNVqdbq9GD9IYjMYTKbouaLXg0Fgrc4bbaVABCqwQnBOHMFl0qNzuDyeyGcbjeH383xC4RocEBwNB4Mh0M8wHFnCyxCo+PIBUKxVKtsdYjCEnidEINE6CAoAG8ybb8C76G6qm0YKg5MgIBNkDQAOKcdrhyPRqweBCuKCCGhVAAqqCU7ILXFyNoKfrLjpojljHjqsb8AF4aFRhL4aAAecbF+QAGU4MEOGbbUGw2FLVar9GIdCS3cLnAA2lAALoAQgdBQAvhQt0A==)

edit: I don't say it's a good solution (I really don't think it's a good one), but it works.

**Reply by jnm2** [reply]
You really do want this to be generic over `(T, T)` and `(T, T, T)` etc, not just `ITuple`, so that type inference works correctly.

**Reply by Captain-Quack** [reply]
Cool feature, by the way.
